test/StripeResource.spec.js

```diff
@@ -32,4 +35,245 @@ describe('StripeResource', function() {
       expect(headers).to.not.include.keys('Stripe-Version');
     });
   });
+
+  describe('Retry Network Requests', function() {
+    // Use a real instance of stripe as we're mocking the http.request responses.
+    var realStripe = require('../lib/stripe')(utils.getUserStripeKey());
+    var options = {
+      host: stripe.getConstant('DEFAULT_HOST'),
+      path: '/v1/charges',
+      data: {
+        amount: 1000,
+        currency: 'usd',
+        source: 'tok_visa',
+        description: 'test'
+      },
+      params: 'amount=1000&currency=usd&source=tok_visa&description=test'
+    };
+
+    afterEach(function() {
+      realStripe.setMaxNetworkRetries(0);
+      stripe.setMaxNetworkRetries(0);
+    });
+
+    after(function() {
+      nock.cleanAll();
+    })
+
+    describe('_request', function() {
+      it('throws an error on connection failure', function(done) {
+        // Mock the connection error.
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff');
+
+        realStripe.charges.create(options.data, function(err) {
+          expect(err.detail.message).to.deep.equal('bad stuff');
+          done();
+        });
+      });
+
+      it('should retry the connection if max retries are set', function(done) {
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff')
+          .post(options.path, options.params)
+          .replyWithError('worse stuff');
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, function(err) {
+          var errorMessage = realStripe.invoices._generateConnectionErrorMessage(1);
+          expect(err.message).to.equal(errorMessage);
+          done();
+        });
+      });
+
+      it('should stop retrying after a successful retry', function(done) {
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff')
+          .post(options.path, options.params)
+          .reply(200, {
+            id: 'ch_123',
+            object: 'charge',
+            amount: 1000,
+          });
+
+        realStripe.setMaxNetworkRetries(2);
```

Comment:
apologies if I don't know `nock` well enough, but is there a guarantee in this test that it did not try to send another request?

Reply:
Once a 200 response is received it should stop there, if it didn't we'd probably see errors like a charge with an idempotency key being called twice.

Reply:
Hmm, I'm not sure nock would give us that functionality... I wanted to double-check that this throws an error: 
```js
// test
nock(url)
  .get(oneThing)
  .get(anotherThing)

// app
get(url); 
get(url); 
get(url); // Nock should raise an UnmockedRequest error, since only two were mocked.
```
My read of [the docs](https://github.com/nock/nock#read-this---about-interceptors) is that this is the behavior, so I think we're good here 😄